src/java/org/apache/cassandra/index/sai/utils/PrimaryKey.java

```diff
@@ -0,0 +1,184 @@
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+
+/**
+ * Representation of the primary key for a row consisting of the {@link DecoratedKey} and
+ * {@link Clustering} associated with a {@link org.apache.cassandra.db.rows.Row}.
+ */
+public class PrimaryKey implements Comparable<PrimaryKey>
+{
+    private final Token token;
+    private final DecoratedKey partitionKey;
+    private final Clustering<?> clustering;
+    private final ClusteringComparator clusteringComparator;
+
+    public PrimaryKey(Token token,
+                      DecoratedKey partitionKey,
+                      Clustering<?> clustering,
+                      ClusteringComparator clusteringComparator)
+    {
+        this.token = token;
+        this.partitionKey = partitionKey;
+        this.clustering = clustering;
+        this.clusteringComparator = clusteringComparator;
+    }
+
+    /**
+     * Returns a {@link PrimaryKeyFactory} for creating {@link PrimaryKey} instances.
+     *
+     * @param clusteringComparator the {@link ClusteringComparator} used by the
+     *                             {@link PrimaryKeyFactory} for clustering comparisons
+     * @return a {@link PrimaryKeyFactory} for {@link PrimaryKey} creation
+     */
+    public static PrimaryKeyFactory factory(ClusteringComparator clusteringComparator)
+    {
+        return new PrimaryKeyFactory(clusteringComparator);
+    }
+
+    /**
+     * Returns the {@link Token} associated with this primary key.
+     *
+     * @return the {@link Token}
+     */
+    public Token token()
+    {
+        return token;
+    }
+
+    /**
+     * Returns the {@link DecoratedKey} associated with this primary key.
+     *
+     * @return the {@link DecoratedKey}
+     */
+    public DecoratedKey partitionKey()
+    {
+        return partitionKey;
+    }
+
+    /**
+     * Returns the {@link Clustering} associated with this primary key
+     *
+     * @return the {@link Clustering}
+     */
+    public Clustering clustering()
+    {
+        return clustering;
+    }
+
+    /**
+     * Returns the {@link PrimaryKey} as a {@link ByteSource} byte comparable representation.
+     *
+     * It is important that these representations are only ever used with byte comparables using
+     * the same elements. This means that {@code asComparableBytes} responses can only be used
+     * together from the same {@link PrimaryKey} implementation.
+     *
+     * @param version the {@link ByteComparable.Version} to use for the implementation
+     * @return the {@code ByteSource} byte comparable.
+     */
+    public ByteSource asComparableBytes(ByteComparable.Version version)
+    {
+        ByteSource tokenComparable = token.asComparableBytes(version);
+        ByteSource keyComparable = partitionKey == null ? null
+                                                        : ByteSource.of(partitionKey.getKey(), version);
```

Comment:
If we have a `null` partition key here, is there any point in continuing to look at the clustering/clustering comparator? It seems like we could just return immediately. If we did that, it seems like we could also get away with not having to pass a non-null comparator to the constructor for the token-only case.